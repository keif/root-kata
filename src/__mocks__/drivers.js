import Driver from "../Driver";

const driverRecordsData =
    `Driver Dan\n` +
    `Driver Alex\n` +
    `Driver Bob\n` +
    `Trip Dan 07:15 07:45 17.3\n` +
    `Trip Dan 06:12 06:32 21.8\n` +
    `Trip Alex 12:01 13:16 42.0`;

const driverRecordsHash = {
    Alex: {
        _driverName: "Alex",
        _averageSpeed: 34,
        _totalMilesDriven: 42,
        _totalDrivingTime: 1.25
    },
    Bob: {
        _driverName: "Bob",
        _averageSpeed: 0,
        _totalMilesDriven: 0,
        _totalDrivingTime: 0
    },
    Dan: {
        _driverName: "Dan",
        _averageSpeed: 47,
        _totalMilesDriven: 39,
        _totalDrivingTime: 0.833333333333333
    }
};

const driveDist = [17.3, 42.0, 0];
const driveTimes = [["07:15", "07:45"], ["06:12", "06:32"], ["12:01", "13:16"]];

const driversNotSorted = [new Driver("Dan"), new Driver("Alex"), new Driver("Bob")];

driversNotSorted.forEach((driver, idx) => {
    driver.setTotalDriveTime(driveTimes[idx]);
    driver.setTotalDrivingDist(driveDist[idx]);
});

const driversSorted = [new Driver("Dan"), new Driver("Bob")];
driversSorted.forEach(driver => {
    let idx = driver._driverName === "Bob" ? 2 : 0;
    driver.setTotalDriveTime(driveTimes[idx]);
    driver.setTotalDrivingDist(driveDist[idx]);
});

const driversSortedArrayDesc = [
    { _driverName: "Alex", _totalDrivingTime: 1.25, _totalMilesDriven: 42, _averageSpeed: 34 },
    { _driverName: "Dan", _totalDrivingTime: 0.833333333333333, _totalMilesDriven: 39, _averageSpeed: 47 },
    { _driverName: "Bob", _totalDrivingTime: 0, _totalMilesDriven: 0, _averageSpeed: 0 }
];

const driversSortedArrayAsc = [
    { _driverName: "Bob", _totalDrivingTime: 0, _totalMilesDriven: 0, _averageSpeed: 0 },
    { _driverName: "Dan", _totalDrivingTime: 0.833333333333333, _totalMilesDriven: 39, _averageSpeed: 47 },
    { _driverName: "Alex", _totalDrivingTime: 1.25, _totalMilesDriven: 42, _averageSpeed: 34 }
];

export {
    driverRecordsData,
    driverRecordsHash,
    driveDist,
    driveTimes,
    driversNotSorted,
    driversSorted,
    driversSortedArrayDesc,
    driversSortedArrayAsc
};
