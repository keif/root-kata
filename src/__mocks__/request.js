import {
    driverRecordsData,
    driversNotSorted,
    driversSorted,
    driversSortedArrayDesc,
    driversSortedArrayAsc
} from "./drivers";

const request = () => {
    return new Promise((resolve, reject) => {
        process.nextTick(() => {
            resolve(driverRecordsData);
        });
    });
};

export { request };
