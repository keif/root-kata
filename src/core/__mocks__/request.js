import {
    driverRecordsData,
    driversNotSorted,
    driversSorted,
    driversSortedArrayDesc,
    driversSortedArrayAsc
} from "../../__mocks__/drivers";

const request = () => {
    return new Promise((resolve, reject) => {
        process.nextTick(() => {
            resolve(driverRecordsData);
        });
    });
};

export { request };
