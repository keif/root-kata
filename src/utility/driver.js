import Driver from "../Driver";
import Trip from "../Trip";

const MIN_SPEED_MPH = 5;
const MAX_SPEED_MPH = 100;

const createDriverHash = fileData => {
    if (!fileData || fileData.length == 0) return {};

    const data = fileData.split("\n");

    return data.reduce(setDriverHash, {});
};

const getDriversRecordsOutput = listOfDrivers => {
    return listOfDrivers.reduce((acc, nxt) => {
        acc += nxt.getRecord() + "\n";
        return acc;
    }, "");
};

const setDriverHash = (driverHash, currentValue) => {
    const [command, name, ...rest] = currentValue.split(" ");

    if (command === "Driver") {
        driverHash[name] = new Driver(name);
    } else if (command === "Trip") {
        driverHash[name].addTrip(...rest);
    } else {
        throw new Error("Unrecognized command");
    }

    return driverHash;
};

const processDrivingRecords = data => {
    const hashTbl = createDriverHash(data);
    const validDriversRecords = validateDriverRecords(hashTbl);
    sortDrivers(validDriversRecords, "dist");

    return getDriversRecordsOutput(validDriversRecords);
};

const sortDrivers = listOfDrivers => {
    return listOfDrivers.sort(
        (currDriver, nextDriver) => nextDriver.getTotalDrivingDist() - currDriver.getTotalDrivingDist()
    );
};

const validateDriverRecords = hashTbl => {
    let idx = 0;
    return Object.keys(hashTbl).reduce((acc, driver) => {
        const currDriver = hashTbl[driver];

        if (currDriver.getAverageSpeed() > -1) {
            acc[idx] = currDriver;
            idx += 1;
        }

        return acc;
    }, []);
};

export {
    createDriverHash,
    getDriversRecordsOutput,
    processDrivingRecords,
    setDriverHash,
    sortDrivers,
    validateDriverRecords
};
