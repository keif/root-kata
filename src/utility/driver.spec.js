import { createDriverHash, getDriversRecordsOutput, sortDrivers, validateDriverRecords } from "./driver";

import { driverRecordsData, driversNotSorted } from "../__mocks__/drivers";

let mockDriversNotSorted = [];

describe("utility/driver.js", () => {
    beforeEach(() => {
        mockDriversNotSorted = driversNotSorted;
    });

    describe("createDriverHash method", () => {
        describe("with no data passed", () => {
            test("returns an empty object", () => {
                expect(createDriverHash()).toEqual({});
            });
        });

        describe("with data passed", () => {
            describe("and the data is valid", () => {
                test("returns a hash map of drivers", () => {
                    expect(createDriverHash(driverRecordsData)).toMatchSnapshot();
                });
            });

            describe("and the data is valid", () => {
                test("throw an error with invalid data", () => {
                    const driverRecordsDataInput = driverRecordsData + "\n";

                    expect(() => createDriverHash(driverRecordsDataInput)).toThrow("Unrecognized command");
                });
            });
        });
    });

    describe("getDriversRecordsOutput method", () => {
        test("returns output of each drivers travel experiences", () => {
            expect(getDriversRecordsOutput(driversNotSorted)).toMatchSnapshot();
        });
    });

    describe("sortDrivers method", () => {
        describe("returns a list of drivers sorted ascending or descending, by distance, miles, or mph", () => {
            test.each([["desc", "dist"], ["asc", "dist"]])("sortDrivers('%s', '%s')", (dir, type) => {
                const hashTbl = createDriverHash(driverRecordsData);
                const validDriversRecords = validateDriverRecords(hashTbl);
                const result = sortDrivers(validDriversRecords, dir, type);

                expect(JSON.parse(JSON.stringify(result))).toMatchSnapshot();
            });
        });
    });

    describe("validateDriverRecords method", () => {
        test("takes a hash table and returns the drivers that meet the criteria", () => {
            expect(validateDriverRecords(mockDriversNotSorted)).toMatchSnapshot();
        });
    });
});
