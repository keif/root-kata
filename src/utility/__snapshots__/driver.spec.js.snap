// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utility/driver.js createDriverHash method with data passed and the data is valid returns a hash map of drivers 1`] = `
Object {
  "Alex": Driver {
    "_averageSpeed": 34,
    "_driverName": "Alex",
    "_totalDrivingTime": 1.25,
    "_totalMilesDriven": 42,
    "_trips": Array [
      Trip {
        "_endTime": "13:16",
        "_startTime": "12:01",
        "_totalMilesDriven": 42,
      },
    ],
  },
  "Bob": Driver {
    "_averageSpeed": 0,
    "_driverName": "Bob",
    "_totalDrivingTime": 0,
    "_totalMilesDriven": 0,
    "_trips": Array [],
  },
  "Dan": Driver {
    "_averageSpeed": 47,
    "_driverName": "Dan",
    "_totalDrivingTime": 0.833333333333333,
    "_totalMilesDriven": 39.1,
    "_trips": Array [
      Trip {
        "_endTime": "07:45",
        "_startTime": "07:15",
        "_totalMilesDriven": 17.3,
      },
      Trip {
        "_endTime": "06:32",
        "_startTime": "06:12",
        "_totalMilesDriven": 21.8,
      },
    ],
  },
}
`;

exports[`utility/driver.js getDriversRecordsOutput method returns output of each drivers travel experiences 1`] = `
"Dan: 17.3 miles @ NaN mph
Alex: 42 miles @ NaN mph
Bob: 0 miles
"
`;

exports[`utility/driver.js sortDrivers method returns a list of drivers sorted ascending or descending, by distance, miles, or mph sortDrivers('asc', 'dist') 1`] = `
Array [
  Object {
    "_averageSpeed": 34,
    "_driverName": "Alex",
    "_totalDrivingTime": 1.25,
    "_totalMilesDriven": 42,
    "_trips": Array [
      Object {
        "_endTime": "13:16",
        "_startTime": "12:01",
        "_totalMilesDriven": 42,
      },
    ],
  },
  Object {
    "_averageSpeed": 47,
    "_driverName": "Dan",
    "_totalDrivingTime": 0.833333333333333,
    "_totalMilesDriven": 39.1,
    "_trips": Array [
      Object {
        "_endTime": "07:45",
        "_startTime": "07:15",
        "_totalMilesDriven": 17.3,
      },
      Object {
        "_endTime": "06:32",
        "_startTime": "06:12",
        "_totalMilesDriven": 21.8,
      },
    ],
  },
  Object {
    "_averageSpeed": 0,
    "_driverName": "Bob",
    "_totalDrivingTime": 0,
    "_totalMilesDriven": 0,
    "_trips": Array [],
  },
]
`;

exports[`utility/driver.js sortDrivers method returns a list of drivers sorted ascending or descending, by distance, miles, or mph sortDrivers('desc', 'dist') 1`] = `
Array [
  Object {
    "_averageSpeed": 34,
    "_driverName": "Alex",
    "_totalDrivingTime": 1.25,
    "_totalMilesDriven": 42,
    "_trips": Array [
      Object {
        "_endTime": "13:16",
        "_startTime": "12:01",
        "_totalMilesDriven": 42,
      },
    ],
  },
  Object {
    "_averageSpeed": 47,
    "_driverName": "Dan",
    "_totalDrivingTime": 0.833333333333333,
    "_totalMilesDriven": 39.1,
    "_trips": Array [
      Object {
        "_endTime": "07:45",
        "_startTime": "07:15",
        "_totalMilesDriven": 17.3,
      },
      Object {
        "_endTime": "06:32",
        "_startTime": "06:12",
        "_totalMilesDriven": 21.8,
      },
    ],
  },
  Object {
    "_averageSpeed": 0,
    "_driverName": "Bob",
    "_totalDrivingTime": 0,
    "_totalMilesDriven": 0,
    "_trips": Array [],
  },
]
`;

exports[`utility/driver.js validateDriverRecords method takes a hash table and returns the drivers that meet the criteria 1`] = `Array []`;
